File: java/memento/History.java
package memento;

import java.util.Stack;

public class History {

    private final Stack<TextEditorState> prevStates;

    public History() {
        this.prevStates = new Stack<>();
    }

    public void saveHistoryState(TextEditorState textEditorState) {
        prevStates.push(textEditorState);
    }

    public TextEditorState undo() {
        if (!prevStates.isEmpty()) {
            return prevStates.pop();
        }
        return null;
    }
}

File: C#/Memento/History.cs
namespace Memento;

/// <summary>
/// Storage for all previous states of the text editor.
/// </summary>
public sealed class History
{
    private readonly Stack<TextEditorState> _prevStates;

    public History()
        => _prevStates = new Stack<TextEditorState>();

    public void SaveHistoryState(TextEditorState textEditorState)
        => _prevStates.Push(textEditorState);

    public TextEditorState? Undo()
        => _prevStates.Count > 0 ? _prevStates.Pop() : null;
}
